Update dependencies from dotnet/roslyn

am11 opened the pull request:
Testing https://github.com/dotnet/runtime/pull/110084 with build fixes.

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241122.2

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-2.24572.2

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241122.15

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-2.24572.15

am11 commented on src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs:
```diff
@@ -739,8 +739,9 @@ private static IEnumerator<T> GetEnumerator(Node? head)
+                T value = current._value;
                 current = current._next;
+                yield return value;
             }
```
@jjonescz, without this change, the test was consistently failing:

```sh
$ ./build.sh libs+libs.tests
$ ./dotnet.sh build -t:Test src/libraries/System.Collections.Concurrent/tests
...
  Starting:    System.Collections.Concurrent.Tests (parallel test collections = on [4 threads], stop on fail = off)
    System.Collections.Concurrent.Tests.ConcurrentStackTests.Enumerator_MoveNext_AfterDisposal(count: 1) [FAIL]
      System.NullReferenceException : Object reference not set to an instance of an object.
      Stack Trace:
        /_/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs(743,0): at System.Collections.Concurrent.ConcurrentStack`1.GetEnumerator(Node head)+MoveNext()
        /_/src/libraries/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs(395,0): at System.Collections.Tests.IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Int32 count)
           at InvokeStub_IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Object, Span`1)
        /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(95,0): at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
    System.Collections.Concurrent.Tests.ConcurrentStackTests.Enumerator_MoveNext_AfterDisposal(count: 75) [FAIL]
...
```
is it expected?

am11 commented on src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs:
```diff
@@ -739,8 +739,9 @@ private static IEnumerator<T> GetEnumerator(Node? head)
-                yield return current._value;
+                T value = current._value;
                 current = current._next;
+                yield return value;
```
@jjonescz, without this change, the tests were consistently failing:
```sh
$ ./build.sh libs+libs.tests
$ ./dotnet.sh build -t:Test src/libraries/System.Collections.Concurrent/tests
...
    Discovering: System.Collections.Concurrent.Tests (method display = ClassAndMethod, method display options = None)
    Discovered:  System.Collections.Concurrent.Tests (found 901 of 909 test cases)
    Starting:    System.Collections.Concurrent.Tests (parallel test collections = on [10 threads], stop on fail = off)
      System.Collections.Concurrent.Tests.ConcurrentStackTests.Enumerator_MoveNext_AfterDisposal(count: 1) [FAIL]
        System.NullReferenceException : Object reference not set to an instance of an object.
        Stack Trace:
          /Users/am11/projects/runtime5/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs(743,0): at System.Collections.Concurrent.ConcurrentStack`1.GetEnumerator(Node head)+MoveNext()
          /Users/am11/projects/runtime5/src/libraries/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs(395,0): at System.Collections.Tests.IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Int32 count)
             at InvokeStub_IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Object, Span`1)
          /Users/am11/projects/runtime5/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(95,0): at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
      System.Collections.Concurrent.Tests.ConcurrentStackTests.Enumerator_MoveNext_AfterDisposal(count: 75) [FAIL]
        System.NullReferenceException : Object reference not set to an instance of an object.
        Stack Trace:
          /Users/am11/projects/runtime5/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs(743,0): at System.Collections.Concurrent.ConcurrentStack`1.GetEnumerator(Node head)+MoveNext()
          /Users/am11/projects/runtime5/src/libraries/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs(395,0): at System.Collections.Tests.IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Int32 count)
             at InvokeStub_IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Object, Span`1)
          /Users/am11/projects/runtime5/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(95,0): at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
    Finished:    System.Collections.Concurrent.Tests
  === TEST EXECUTION SUMMARY ===
     System.Collections.Concurrent.Tests  Total: 2460, Errors: 0, Failed: 2, Skipped: 0, Time: 74,172s
...
```
is it expected?

jjonescz replied:
@jcouv is this due to https://github.com/dotnet/roslyn/pull/75908?

jcouv replied:
Yes, that is the cause.

Here's a [sharplab repro](https://sharplab.io/#v2:C4LgTgrgdgPgsAKAPRIAQCECmALAhgNwEsB7MVAI0wDNTNUBjPKAczoApCpUAiAE2twQANsG4UwuKIwCUIVAAcwnYAGceAFUiYALAGZUvQivnEVmXqlxVgmMpoiZuiFBhwESZKzbKNJrVBxc3AC2uJxi5BJS2LKoAHIASgCizmhJAB7ymPQ2FpR4RKRyispq3PY6+obGpuaoAGK4QmZOCIj4uGQdQg5qALyoUJgA7qgAwgA8ygB8bADa2gBMADSoegC60gDc7Z2omFAQwba4wKSoA929AHQA4pjASYfHEmdgbNuIAAIAjNdjxCgKmIQkw1wA6kobGwDkcTm9rgBZYj4TBxTDpYAfT4IX7/QHA0EQqGYGHPeGkf4QMBgA7AHGIWEvU6UgAiRhMZg+O1xfwBQJBYMhhGhYmqnLq3AZvPxAqJwuhTIpYCRKLRGKx0ml30W4wm6mmiAA3ohUGaFEoOjZUGYmnUvrq4sR+KbzSaEObPahlLYoE1ULTcLxAUIAJ6odSoAD6V0wW1QLgAak0HKhiFRUMBsHQoM6wa6vd6oN4/UJ4nno0NMfGXOjMQpiD6VYgC16m6Xy/w2JHY9JW573YXCzGU3RLqOeUOvVGq8ALoNhEJJ0OAL6ttdtD3mkpWuhO/jR7NB+Mtrdmr76MZsX7aa6/XT66YMEGgnIkKB9s+oQdD/d0IS4CowB/vOhxCEu/bmjQgaMAEkaYKCxzFkWz7gdkwDvp+U4/lOZogUMwwgQMBGdqSCGYEh9LLrhgwjH+1wzhq84AUBf7UbhLHARWxF0Xm7GehukFmlGR4WAMnFseup6eheqA3tcACSTxwq8pCPqg9yPOSqnvFhbpCXJADsGkPMpzJvGwImYEGOICdJ26WqcdC/AAbHJPy3kp2ksmA6maWZypsCBol6WaOFeiB9DUrSyEDKJ/HmoQGZsFFNJ0t6ahgUI0jubKhJCiSbBiBR8jAOGUo8gZwzYIQoIBKlMVzgAhMRi6hV64VTr8PxGQw0V0gxsYJYWDXpQMo3Fgxs7DV6eL8vlxIiqS3BeLYPC2auUkICuQA==).

In the "main" branch, we NRE.
In the "default" branch (ie. before the change to clear locals during disposal), we printed a result without crash. But it looks like that result was incorrect already... 
We were incorrectly allowing enumeration to continue. 

The correct behavior would be for disposal to set the enumerator state to "after" (thereby causing `MoveNext` to return `false` thereafter).
This issue becomes more observable after the change to clear local state (NRE in the present scenario).
Filed [https://github.com/dotnet/roslyn/issues/76078](https://github.com/dotnet/roslyn/issues/76078)

I will discuss with @AlekseyTs @333fred on how best to proceed. I thinking to revert the change temporarily and fix the disposal bug first.

am11 replied:
@jcouv, thanks for the info. I think we can wait, since there are some other NRE failures in CI which appear to be related.

jcouv replied:
That works, thanks. 
FWIW, here's the [fix in progress](https://github.com/dotnet/roslyn/pull/76090)

jcouv replied:
@am11 The fix was merged into roslyn.

am11 replied:
@jcouv, thanks for the update. I have updated to latest version. However, this test is still failing without faf8f96b2861b656f905768c054cc87c64d09d62. Were we expecting it to work without that change in the consumer code?

am11 replied:
@jcouv, thanks for the update. I have updated to latest version. However, `System.Collections.Concurrent.Tests` are still failing without faf8f96b2861b656f905768c054cc87c64d09d62. Were we expecting it to work without that change in the consumer code?

am11 replied:
It was still the old version. Now it's the one with your fix. It worked! 👍

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241124.5

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-2.24574.5

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241125.18

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-3.24575.18

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241127.2

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-3.24577.2

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241128.2

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-3.24578.2

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241128.3

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-3.24578.3

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241130.1

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-3.24580.1

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241130.2

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-3.24580.2

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241201.2

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-3.24601.2

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/roslyn build 20241202.26

Microsoft.SourceBuild.Intermediate.roslyn , Microsoft.CodeAnalysis , Microsoft.CodeAnalysis.CSharp , Microsoft.Net.Compilers.Toolset
 From Version 4.13.0-2.24570.4 -> To Version 4.13.0-3.24602.26

am11 committed:
Use explicit AsSpan

am11 committed:
Merge branch 'main' into darc-main-19369168-1f4a-490d-90eb-b9e754971a1a

am11 committed:
Fix LibraryImportGenerator.UnitTests

am11 committed:
Fix regex tests

am11 commented:
@dotnet/illink, @MichalStrehovsky, any thoughts about the last error under Build Analysis https://github.com/dotnet/runtime/pull/110105/checks?check_run_id=33908815428:

```
    ❌ Mono.Linker.Tests.TestCases.All.DataFlow(t: \"CompilerGeneratedCodeAccessedViaReflection\")
    Exception Message

        Unexpected warning found: ILC: Trim analysis warning IL2062: Mono.Linker.Tests.Cases.DataFlow.CompilerGeneratedCodeAccessedViaReflection.IteratorStateMachines.<IteratorWithIntegerDataflow>d__3.MoveNext(): Value passed to parameter 'type' of method 'Mono.Linker.Tests.Cases.Expectations.Helpers.DataFlowTypeExtensions.RequiresPublicMethods(Type)' can not be statically determined and may not meet 'DynamicallyAccessedMembersAttribute' requirements.

    Stack Trace

           at Mono.Linker.Tests.TestCasesRunner.ResultChecker.VerifyLoggedMessages(AssemblyDefinition original, TrimmingTestLogger logger, Boolean checkRemainingErrors) in /_/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCasesRunner/ResultChecker.cs:line 405
           at Mono.Linker.Tests.TestCasesRunner.ResultChecker.AdditionalChecking(TrimmedTestCaseResult linkResult, AssemblyDefinition original) in /_/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCasesRunner/ResultChecker.cs:line 157
           at Mono.Linker.Tests.TestCasesRunner.ResultChecker.Check(TrimmedTestCaseResult testResult) in /_/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCasesRunner/ResultChecker.cs:line 70
           at Mono.Linker.Tests.TestCases.All.Run(String testName) in /_/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCases/TestSuites.cs:line 157
           at Mono.Linker.Tests.TestCases.All.DataFlow(String t) in /_/src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCases/TestSuites.cs:line 16
           at InvokeStub_All.DataFlow(Object, Span`1)
           at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

sbomer committed:
Adjust trimming test for new codegen

With this roslyn update, the generated iterator state machine
for this test includes a Dispose method which assigns null to the
array, inhibiting dataflow analysis.

am11 committed:
Merge branch 'main' into darc-main-19369168-1f4a-490d-90eb-b9e754971a1a

am11 commented:
Thanks @sbomer for pushing the fix!

am11 commented:
@akoeplinger @ViktorHofer PTAL.

sbomer commented:
/backport to release/9.0-staging
